src/llmq/quorums_dkgsessionhandler.cpp

```diff
@@ -0,0 +1,577 @@
+{
+    auto msgs = pendingMessages.PopAndDeserializeMessages<Message>(maxCount);
+    if (msgs.empty()) {
+        return false;
+    }
+
+    std::vector<uint256> hashes;
+    std::vector<std::pair<NodeId, std::shared_ptr<Message>>> preverifiedMessages;
+    hashes.reserve(msgs.size());
+    preverifiedMessages.reserve(msgs.size());
+
+    for (const auto& p : msgs) {
+        if (!p.second) {
+            LogPrint(BCLog::NET, "%s -- failed to deserialize message, peer=%d\n", __func__, p.first);
+            {
+                LOCK(cs_main);
+                Misbehaving(p.first, 100);
+            }
+            continue;
+        }
+        const auto& msg = *p.second;
+
+        auto hash = ::SerializeHash(msg);
+
+        bool ban = false;
+        if (!session.PreVerifyMessage(hash, msg, ban)) {
+            if (ban) {
+                LogPrint(BCLog::NET, "%s -- banning node due to failed preverification, peer=%d\n", __func__, p.first);
+                {
+                    LOCK(cs_main);
+                    Misbehaving(p.first, 100);
+                }
+            }
+            LogPrint(BCLog::NET, "%s -- skipping message due to failed preverification, peer=%d\n", __func__, p.first);
+            continue;
+        }
+        hashes.emplace_back(hash);
+        preverifiedMessages.emplace_back(p);
+    }
+    if (preverifiedMessages.empty()) {
+        return true;
+    }
+
+    auto badNodes = BatchVerifyMessageSigs(session, preverifiedMessages);
+    if (!badNodes.empty()) {
+        LOCK(cs_main);
+        for (auto nodeId : badNodes) {
+            LogPrint(BCLog::NET, "%s -- failed to verify signature, peer=%d\n", __func__, nodeId);
+            Misbehaving(nodeId, 100);
+        }
+    }
+
+    for (size_t i = 0; i < preverifiedMessages.size(); i++) {
+        NodeId nodeId = preverifiedMessages[i].first;
+        if (badNodes.count(nodeId)) {
+            continue;
+        }
+        const auto& msg = *preverifiedMessages[i].second;
+        bool ban = false;
+        session.ReceiveMessage(hashes[i], msg, ban);
+        if (ban) {
+            LogPrint(BCLog::NET, "%s -- banning node after ReceiveMessage failed, peer=%d\n", __func__, nodeId);
+            LOCK(cs_main);
+            Misbehaving(nodeId, 100);
+            badNodes.emplace(nodeId);
+        }
+    }
+
+    return true;
+}
+
+void CDKGSessionHandler::HandleDKGRound()
+{
+    uint256 curQuorumHash;
+
+    WaitForNextPhase(QuorumPhase_None, QuorumPhase_Initialized, UINT256_ZERO, []{return false;});
+
+    {
+        LOCK(cs);
+        pendingContributions.Clear();
+        pendingComplaints.Clear();
+        pendingJustifications.Clear();
+        pendingPrematureCommitments.Clear();
+        curQuorumHash = quorumHash;
+    }
+
+    const CBlockIndex* pindexQuorum{nullptr};
+    {
+        LOCK(cs_main);
+        if (mapBlockIndex.count(curQuorumHash))
+            pindexQuorum = mapBlockIndex.at(curQuorumHash);
+    }
+    if (!pindexQuorum) {
```

Comment:
Can simplify it using the recently added `LookupBlockIndex` function

Reply:
updated (also in `CDKGSessionHandler::HandleDKGRound()`)